Cherry Picker and Do Not Resell Consent APIs for Azure Communication Services Phone Numbers service

danielortega-msft opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

danielortega-msft committed:
Add cherry picker changes

danielortega-msft committed:
Minor fixes for cherry picker API

danielortega-msft committed:
Make assignment type optional for browse

danielortega-msft committed:
Fix examples

danielortega-msft committed:
fix prettier

danielortega-msft committed:
fix examples formatting

danielortega-msft commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -1444,6 +1875,26 @@
         "cost": {
           "$ref": "#/definitions/PhoneNumberCost"
+        },
+        "operatorId": {
+          "description": "Id of the operator that provided the number",
+          "type": "string"
+        },
+        "operatorName": {
+          "description": "Name of the operator that provided the number",
+          "type": "string"
+        },
+        "phoneNumberSource": {
+          "description": "Source of the number, e.g. Cloud or OperatorConnect",
+          "enum": [
+            "cloud",
+            "operatorConnect"
+          ],
+          "type": "string",
+          "x-ms-enum": {
+            "name": "PhoneNumberSource",
+            "modelAsString": true
+          }
```
These properties are added from an existing preview API `2022-06-01-preview`: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/communication/data-plane/PhoneNumbers/preview/2022-06-01-preview

They appear in the diff because they weren't present in the preview API this is based upon, `2023-05-01-preview`.  The only change made is that they weren't marked as required to avoid breaking changes; which should be fine because they are actually readonly (I didn't add this to the properties because it wasn't in the original spec).

danielortega-msft committed:
Update operation name for reservation purchase

danielortega-msft committed:
fix examples and parameter names for cherry picker

danielortega-msft committed:
change add/remove from reservation operations to merge patch

danielortega-msft committed:
Fix create reservation example

danielortega-msft committed:
run prettier

danielortega-msft committed:
fix readme.md file for phone numbers

danielortega-msft committed:
Fix conflict in CommunicationErrorResponse version

danielortega-msft committed:
Add missing type properties

danielortega-msft committed:
Create specific type for capabilities in browse request

This prevents coupling between the update capabilities logic and the browse request logic

danielortega-msft committed:
fix typo

danielortega-msft committed:
Update AvailablePhoneNumber model to handle bulk updates

danielortega-msft committed:
Add failed status to reservation

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -427,6 +427,59 @@
         }
       }
     },
+    "/availablePhoneNumbers/countries/{countryCode}/:browse": {
```
`POST` actions should not use a slash:

```suggestion
    "/availablePhoneNumbers/countries/{countryCode}:browse": {
```

If you already have `POST` actions that do this, however, feel free to ignore. You should be consistent within an API spec.

danielortega-msft replied:
Our existing API already has the `/:` format for existing POST actions.

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -427,6 +427,59 @@
+          {
+            "in": "path",
+            "name": "countryCode",
+            "description": "The ISO 3166-2 country code, e.g. US.",
```
Is this truly upper case, or are you case-insensitive? HTTP recommends case-sensitive paths, and I'm used to seeing e.g., "us" not "US", hence the question.

danielortega-msft replied:
This is case insensitive, which is the same behavior for our GA APIs as well. That said, I do think it makes sense for both "US" and "us" to be valid ways to identify a country.

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -530,6 +583,325 @@
+              "$ref": "#/definitions/PhoneNumbersReservation"
+            }
+          },
+          "409": {
```
Do not document other errors unless novel. 409 should be expected and would err regardless.

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -530,6 +583,325 @@
+          }
+        }
+      },
+      "patch": {
```
Why `PUT` and `PATCH`? Generally, we recommend `PATCH` for create-or-update semantics. See https://aka.ms/azapi/guidelines

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -530,6 +583,325 @@
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
```
Nit: Second time I've seen this same parameter in just the diff. I imagine it's duplicated more. You should consider refactoring common parameters into the `parameters` section of the swagger.

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -530,6 +583,325 @@
+        }
+      }
+    },
+    "/availablePhoneNumbers/reservations/{reservationId}/:purchase": {
```
Same comment about no slash for a `POST` action unless the API has already shipped that way in the past for other endpoints.

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -1238,6 +1611,10 @@
           "description": "The search id.",
           "type": "string"
+        },
+        "consentToNotResellNumbers": {
```
This reads very confusingly. They consent NOT to resell? Seems they should consent to resell, thus defaulting to false for privacy. Regardless of privacy defaults, just the wording alone is very confusing. What would a UI display string look like?

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -1467,6 +1864,270 @@
+      ],
+      "type": "object",
+      "properties": {
+        "phoneNumberType": {
```
We recommend "kind" over "type" for discriminators. "Type" can cause conflicts and confusion in various SDK languages.

danielortega-msft replied:
This is reusing a GA'd definition.

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -1467,6 +1864,270 @@
+        "capabilities": {
+          "$ref": "#/definitions/PhoneNumberBrowseCapabilitiesRequest"
+        },
+        "assignmentType": {
```
Same here, or anywhere you use "type".

danielortega-msft replied:
Same here, this is a GA'd definition.

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -1467,6 +1864,270 @@
+        "calling": {
+          "description": "Capability value for calling.",
+          "enum": [
+            "none",
```
Rather than "none" values, you should treat the absence of these various enum types (like `sms` below) as "none".

heaths reviewed (approved):
I see nothing blocking for a preview.

danielortega-msft committed:
Address ARB feedback

- Fixes endpoint parameter in examples
- Replaces PUT+PATCH in favor of PATCH only for creating or updating reservations
- Removes Reservation PartiallyCompleted and Failed status
- Create separate definitions for ReservationStatus and AvailablePhoneNumberStatus
- Create separate definitions for duplicate parameters

danielortega-msft committed:
Declare ReservationStatus and AvailablePhoneNumberStatus inline

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/examples/CreateOrUpdateReservation.json:
```diff
@@ -18,6 +18,32 @@
     }
   },
   "responses": {
+    "201": {
```
Why is this returning a 200 and 201? This is suspect.

danielortega-msft replied:
That's a 201 for newly created reservations, 200 for updates on existing ones. I took this pattern from the guidelines:
https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#create--update--replace-processing-rules

heaths commented on specification/communication/data-plane/PhoneNumbers/preview/2023-10-01-preview/phonenumbers.json:
```diff
@@ -640,36 +640,27 @@
-            "description": "The id of the reservation.",
-            "required": true,
-            "type": "string"
+            "$ref": "#/parameters/ReservationIdParameter"
```
The intent was to `$ref` these parameters everywhere used - not just in this one place; otherwise, there's not much point. This saves you typing, but also from definitions accidentally diverging, which is typically unintentional.

danielortega-msft replied:
I'm not sure I get this comment. For every API using the ReservationId or Reservation as parameters, they are now using the `$ref` instead of declaring them inline.

I didn't change any of the already existing APIs, though; is that what you are suggesting?

heaths replied:
Yes. There should be no breaking changes. It's more of a nit, though.

danielortega-msft committed:
Consolidate Query parameters for phone numbers API
Remove readonly properties from required list

danielortega-msft commented:
Abandoning in favor of: https://github.com/Azure/azure-rest-api-specs/pull/28670
